### PR TITLE
Changed random number generation to use the most recent numpy API

### LINKED ISSUE #69: Ensure reproducibility of image production runs
Reproduce results obtained by the various BTK steps would be a real added value.

So far the introductory notebook makes use of `numpy.random.seed()` but because many generators are involved, we should probably use something more scalable like passing around a random generator like `numpy.random.RandomState()`.

### PR DESCRIPTION
**Pull request by thuiop**
I migrated the BTK to use the newest numpy random number generation API. However, this causes an issue with the multiprocessing ; it no longer returns the same thing as the non multiprocessing case (I am not sure how it managed to do that in the first place actually). It is still reproducible, even for the multiprocessing case (with a set number of CPUs). Hopefully, this should make the simulations completely reproducible as long as the multiprocessing is the same. What are your thoughts on this ?

Closes #69.

### CODE REVIEW

**Review by ismael-mendoza — APPROVED** [review]
Thanks Thomas this looks great, just a couple of small questions below. 

I think it's ok that using different number of CPUs does not give the same result. As long as you can reproduce results by using the same number of CPUs we should be good. Feel free to remove the corresponding unit test (that's the only one failing right?)

File: btk/sampling_functions.py
Comment (thuiop) [reply btk/sampling_functions.py:197]: No, the ones with the 0 use the seed 0. They should only be in the tests.

File: btk/draw_blends.py
Comment (thuiop) [reply btk/draw_blends.py:420]: Yes, it will give the same seed to the Galsim one everytime so it should be fine.

File: btk/sampling_functions.py
Comment (ismael-mendoza) [reply btk/sampling_functions.py:197]: gotcha

**Comment by aboucaud** [comment]
Thanks @thuiop for taking on this task but IMHO this PR was merged a bit too quickly (<24h) for such an important step. 
Random generation is the fondation of reproducibility so I think some of the choices made here should require more discussions before being merged to `main`. 
I serve mostly as a BTK reviewer so I'm just sharing my opinion. As maintainers, this is your call of course.

**Comment by thuiop** [comment]
You may be right. What comments did you have in mind ?

**Comment by ismael-mendoza** [comment]
Hi @aboucaud - thanks for your comment. We are definitely happy to discuss this more and revert/make changes as necessary. Feel free to let us know additional things we might want to discuss in a github issue or during the meeting on Wednesday. Thanks!

File: btk/draw_blends.py
Comment (aboucaud) [inline btk/draw_blends.py:133]: here should be `seed` and not `rng`

File: btk/draw_blends.py
Comment (aboucaud) [inline btk/draw_blends.py:196]: so here if I give `rng = numpy.random` this will work but good luck for reproducibility.
I guess we should restrict only the cases to
- a seed is provided
- a default seed is used

File: btk/draw_blends.py
Comment (aboucaud) [reply btk/draw_blends.py:420]: For clarity, please separate the logic in two lines 

```python
galsim_seed = self.rng.integers(xxx)
galsim.random.BaseDeviate(seed=galsim_seed)
```

But I am still not convinced that this will always return the same seed. It will depend on the order of the computations. If there is a `if` statement somewhere before that calls to `self.rng` this will return a different value for the seed.

That's why I think we need to be careful here because the way you tested it might not be the only way it will be used.

File: btk/sampling_functions.py
Comment (aboucaud) [inline btk/sampling_functions.py:58]: Having `default_rng()` without specified seed as the default value of a keyword does not seem a good idea to me.
I'd rather have `None` and then do something about it.

File: tests/test_draw.py
Comment (aboucaud) [inline tests/test_draw.py:17]: here the seed is set to 0 without justification.

File: tests/test_draw.py
Comment (aboucaud) [inline tests/test_draw.py:60]: Here again seed set to 0

File: tests/test_draw.py
Comment (aboucaud) [reply tests/test_draw.py:17]: my bad, I get it, it's for testing.
Still it may be a good idea to have the test seed apparent somewhere

```python
TEST_SEED = 0
...
rng = np.random.default_rng(TEST_SEED)
```

**Comment by aboucaud** [comment]
Here are some of the comments I had.

Concerning the multiprocessing, it might be valuable to have a look at
https://albertcthomas.github.io/good-practices-random-number-generators/

I hope I'll be able to make it to the telecon tomorrow.

File: btk/draw_blends.py
Comment (thuiop) [reply btk/draw_blends.py:133]: Why ? It is not necessary a seed, it can directly be a `numpy.random.Generator`.

File: btk/draw_blends.py
Comment (thuiop) [reply btk/draw_blends.py:420]: Why would there be an additional `if` statement if we use the same parameters for the generation ?

File: tests/test_draw.py
Comment (thuiop) [reply tests/test_draw.py:17]: This can be a good idea.

**Comment by ismael-mendoza** [comment]
Another example from Axel: https://github.com/LSSTDESC/descwl-shear-sims/blob/master/shear_meas_tests/test_shear_meas.py

**Comment by thuiop** [comment]
> Another example from Axel: https://github.com/LSSTDESC/descwl-shear-sims/blob/master/shear_meas_tests/test_shear_meas.py

This one uses RandomState, which is from the old numpy API (which is still working but we should not start using it now).

**Pull request merged**
